src/Confluent.Kafka/Consumer.cs

```diff
@@ -110,863 +246,400 @@ public class Consumer<TKey, TValue> : IDisposable
+        /// </returns>
         /// <remarks>
-        ///     Executes on the same thread as every other Consumer event handler (except OnLog which may be called from an arbitrary thread).
+        ///     OnPartitionsAssigned/Revoked and OnOffsetsCommitted events
+        ///     may be invoked as a side-effect of calling this method (on 
+        ///     the same thread).
         /// </remarks>
-        public event EventHandler<List<TopicPartition>> OnPartitionsRevoked;
+        public ConsumeResult<TKey, TValue> Consume(TimeSpan timeout)
+            => Consume(timeout.TotalMillisecondsAsInt());
 
-        /// <summary>
-        ///     Raised to report the result of (automatic) offset commits.
-        ///     Not raised as a result of the use of the CommitAsync method.
-        /// </summary>
-        /// <remarks>
-        ///     Executes on the same thread as every other Consumer event handler (except OnLog which may be called from an arbitrary thread).
-        /// </remarks>
-        public event EventHandler<CommittedOffsets> OnOffsetsCommitted;
 
-        /// <summary>
-        ///     Raised on critical errors, e.g. connection failures or all 
-        ///     brokers down. Note that the client will try to automatically 
-        ///     recover from errors - these errors should be seen as 
-        ///     informational rather than catastrophic
-        /// </summary>
-        /// <remarks>
-        ///     Executes on the same thread as every other Consumer event handler (except OnLog which may be called from an arbitrary thread).
-        /// </remarks>
-        public event EventHandler<Error> OnError;
 
         /// <summary>
-        ///     Raised when a consumed message has an error != NoError (both when Consume or Poll is used for polling).
+        ///     Poll for new messages / events. Blocks until a consume result
+        ///     is available or the operation has been cancelled.
         /// </summary>
+        /// <param name="cancellationToken">
+        ///     A cancellation token that can be used to cancel this operation.
+        /// </param>
+        /// <returns>
+        ///     The consume result.
+        /// </returns>
         /// <remarks>
-        ///     Executes on the same thread as every other Consumer event handler (except OnLog which may be called from an arbitrary thread).
+        ///     OnPartitionsAssigned/Revoked and OnOffsetsCommitted events
+        ///     may be invoked as a side-effect of calling this method (on 
+        ///     the same thread).
         /// </remarks>
-        public event EventHandler<Message> OnConsumeError;
+        public ConsumeResult<TKey, TValue> Consume(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var result = Consume(100);
+                if (result.Message == null) { continue; }
```

Comment:
Hi! I'm testing v1.0-experiental-11, and I got `NullReferenceException` when `internal ConsumeResult<TKey, TValue> Consume(int)` returns null.
Maybe should have a null check here (`result?.Message`).

Reply:
thanks for reporting - missed that, fixed.